decode_test.go

```diff
@@ -1296,35 +1213,14 @@ func TestUnmarshalerTypeError(t *testing.T) {
 		"  line 1: foo\n" +
 		"  line 1: bar\n" +
 		"  line 1: cannot unmarshal !!str `B` into int"
-	if err == nil {
-		t.Fatalf("Unmarshal() expected error, got none.")
-	}
-	matched, mre := regexp.MatchString(expectedError, err.Error())
-	if mre != nil {
-		t.Fatalf("regexp.MatchString() returned error: %v", mre)
-	}
-	if !matched {
-		t.Fatalf("Unmarshal() returned error %q, want match for %q", err, expectedError)
-	}
-	if v.M["abc"] == nil {
-		t.Fatalf("v.M[\"abc\"] is nil, want non-nil")
-	}
-	if v.M["def"] != nil {
-		t.Fatalf("v.M[\"def\"] is not nil, want nil")
-	}
-	if v.M["ghi"] == nil {
-		t.Fatalf("v.M[\"ghi\"] is nil, want non-nil")
-	}
-	if v.M["jkl"] != nil {
-		t.Fatalf("v.M[\"jkl\"] is not nil, want nil")
-	}
-
-	if !reflect.DeepEqual(v.M["abc"].value, 1) {
-		t.Fatalf("v.M[\"abc\"].value is %v, want %v", v.M["abc"].value, 1)
-	}
-	if !reflect.DeepEqual(v.M["ghi"].value, 3) {
-		t.Fatalf("v.M[\"ghi\"].value is %v, want %v", v.M["ghi"].value, 3)
-	}
+	testutil.AssertErrorMatches(t, err, expectedError)
+	testutil.AssertNotNil(t, v.M["abc"])
+	testutil.AssertIsNil(t, v.M["def"])
+	testutil.AssertNotNil(t, v.M["ghi"])
+	testutil.AssertIsNil(t, v.M["jkl"])
+
+	testutil.AssertEqual(t, v.M["abc"].value, 1)
```

Comment:
You could have used AssertDeepEqual as in the code you replace.

But then I wonder how the code would suffer from having AssertEqual using reflect.DeepEqual and having no AssertDeepEqual.

It would be slower, and more restrictive. But I feel like it could be OK, and would avoid to bump Go minimal version